src/memory-analyzer/gdb_api.cpp

```diff
@@ -0,0 +1,262 @@
+// Copyright 2018 Author: Malte Mues <mail.mues@gmail.com>
+#ifdef __linux__
+#include <cctype>
+#include <cerrno>
+#include <cstdio>
+#include <cstring>
+#include <regex>
+
+#include "gdb_api.h"
+#include <goto-programs/goto_model.h>
+
+gdb_apit::gdb_apit(const char *arg)
+  : binary_name(arg), buffer_position(0), last_read_size(0)
+{
+  memset(buffer, 0, MAX_READ_SIZE_GDB_BUFFER);
+}
+
+int gdb_apit::create_gdb_process()
+{
+  if(pipe(pipe_input) == -1)
+  {
+    throw gdb_interaction_exceptiont("could not create pipe for stdin!");
+  }
+  if(pipe(pipe_output) == -1)
+  {
+    throw gdb_interaction_exceptiont("could not create pipe for stdout!");
+  }
+
+  gdb_process = fork();
+  if(gdb_process == -1)
+  {
+    throw gdb_interaction_exceptiont("could not create gdb process.");
+  }
+  if(gdb_process == 0)
+  {
+    // child process
+    close(pipe_input[1]);
+    close(pipe_output[0]);
+    dup2(pipe_input[0], STDIN_FILENO);
+    dup2(pipe_output[1], STDOUT_FILENO);
+    dup2(pipe_output[1], STDERR_FILENO);
+    dprintf(pipe_output[1], "binary name: %s\n", binary_name);
+    char *arg[] = {
+      const_cast<char *>("gdb"), const_cast<char *>(binary_name), NULL};
+
+    dprintf(pipe_output[1], "Loading gdb...\n");
+    execvp("gdb", arg);
+
+    // Only reachable, if execvp failed
+    int errno_value = errno;
+    dprintf(pipe_output[1], "errno in child: %s\n", strerror(errno_value));
+  }
+  else
+  {
+    // parent process
+    close(pipe_input[0]);
+    close(pipe_output[1]);
+    write_to_gdb("set max-value-size unlimited\n");
+  }
+  return 0;
+}
+
+void gdb_apit::terminate_gdb_process()
+{
+  close(pipe_input[0]);
+  close(pipe_input[1]);
+}
+
+void gdb_apit::write_to_gdb(const std::string &command)
+{
+  if(
+    write(pipe_input[1], command.c_str(), command.length()) !=
+    (ssize_t)command.length())
+  {
+    throw gdb_interaction_exceptiont("Could not write a command to gdb");
+  }
+}
+
+std::string gdb_apit::read_next_line()
+{
+  char token;
+  std::string line;
+  do
+  {
+    while(buffer_position >= last_read_size)
+    {
+      read_next_buffer_chunc();
+    }
+    token = buffer[buffer_position];
+    line += token;
+    ++buffer_position;
+  } while(token != '\n');
+  return line;
+}
+
+void gdb_apit::read_next_buffer_chunc()
+{
+  memset(buffer, 0, last_read_size);
+  const auto read_size =
+    read(pipe_output[0], &buffer, MAX_READ_SIZE_GDB_BUFFER);
```

Comment:
I'm not personally offended by this kind of thing but it isn't very C++  Would it be possible to use the standard I/O streams to avoid needing to do our own line buffering?

Reply:
Yes that is right, but I haven't found a really nice way, to read and write to a child process from C++. Therefore the fork and pipe approach is mainly borrowed from c, but as far as I have seen, there is no standard way, to open a stream buffer on a file descriptor. At least, I haven't found a buffer in I/O stream that opens a file descriptor and searching for this on the internet also hasn't revealed something that works well. For sure there are some Boost libraries doing this for you, but I don't want this dependency.
Instead of reading into this fixed size buffer, I might build some class around the file descriptor and then implement ```>>``` and ```get``` on them. I would then only use ```get```, as I care about whitespaces.

Sure, I might read only a single byte, so I don't maintain an internal buffer, but I see a performance issue here.

I am open for further suggestions and willed to learn, how to change this, but I don't have another idea at hand. Writing my own class with the methods required, doesn't seem to have a benefit from my point of view instead of using this buffer.